### ISSUE TITLE
[local_auth] authenticate() crashes the app on iOS when localizedReason is empty

### ISSUE THREAD

**Issue by enricobenedos** [flutter/flutter#77794]
## Steps to Reproduce

1. Run `flutter create bug`.
2. Update the files as follows:
```dart
import 'package:flutter/material.dart';
import 'package:local_auth/local_auth.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key key, this.title}) : super(key: key);

  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: Center(
        child: TextButton(
          child: Text('Start biometric auth'),
          onPressed: () async {
            final LocalAuthentication localAuth = LocalAuthentication();
            final bool canCheckBiometrics = await localAuth.canCheckBiometrics;

            if (canCheckBiometrics) {
              final List<BiometricType> availableBiometrics =
                  await localAuth.getAvailableBiometrics();
              if (availableBiometrics.length > 0) {
                print('Waiting for biometric authentication.');

                bool isAuthenticated;
                try {
                  isAuthenticated = await localAuth.authenticate(
                      localizedReason: ''
                      // localizedReason: I18n.of(context).login_biometricAuthMsg,
                      );
                } catch (e) {
                  print(e);
                  return;
                }
                if (isAuthenticated) {
                  print('Authenticated!!');
                } else {
                  print('Authentication failed!!');
                }
              }
            } else {
              print('Cannot check biometrics');
            }
          },
        ),
      ),
    );
  }
}
```
3. Add `local_auth: ^1.1.0` to `pubspec.yaml`
4. Add the key to `Info.plist` file
```xml
<key>NSFaceIDUsageDescription</key>
<string>Why is my app authenticating using face id?</string>
```

**Expected results:** 
I expect that the iOS simulator show me the FaceId/TouchId auth form, then I can emulate the face/finger recognition using the simulator tools.

**Actual results:** 
When the app call `authenticate()` method completely crash with no info about the problem.

<details>
  <summary>Logs</summary>

[verbose_log.txt](https://github.com/flutter/flutter/files/6115279/verbose_log.txt)

```
Analyzing local_auth_bug...                                             
No issues found! (ran in 1.8s)
```

```
[✓] Flutter (Channel stable, 2.0.1, on macOS 11.2.2 20D80 darwin-x64, locale en-GB)
    • Flutter version 2.0.1 at /Users/develop/SDKs/flutter
    • Framework revision c5a4b4029c (6 days ago), 2021-03-04 09:47:48 -0800
    • Engine revision 40441def69
    • Dart version 2.12.0

[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.2)
    • Android SDK at /Users/develop/Library/Android/sdk
    • Platform android-30, build-tools 30.0.2
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6915495)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 12.4, Build version 12D4e
    • CocoaPods version 1.10.1

[✗] Chrome - develop for the web (Cannot find Chrome executable at /Applications/Google Chrome.app/Contents/MacOS/Google
    Chrome)
    ! Cannot find Chrome. Try setting CHROME_EXECUTABLE to a Chrome executable.

[✓] Android Studio (version 4.1)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6915495)

[✓] VS Code (version 1.54.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.20.0

[✓] Connected device (1 available)
    • iPhone 12 Pro Max (mobile) • 3F2E6288-280C-45EF-BCA4-1236F663267E • ios • com.apple.CoreSimulator.SimRuntime.iOS-14-4
      (simulator)

! Doctor found issues in 1 category.
```

</details>

**Comment by enricobenedos** [comment]
With a real device the verbose log is clear:
```
[   +1 ms] Flutter run key commands.
[   +1 ms] r Hot reload. 🔥🔥🔥
[        ] R Hot restart.
[        ] h Repeat this help message.
[        ] d Detach (terminate "flutter run" but leave application running).
[        ] c Clear the screen
[        ] q Quit (terminate the application on the device).
[        ] An Observatory debugger and profiler on iPad 2018 Develop (3662) is available at: http://127.0.0.1:59434/jfRiHv2G7PA=/
[        ] 
           Flutter DevTools, a Flutter debugger and profiler, on iPad 2018 Develop (3662) is available at:
http://127.0.0.1:9106?uri=http%3A%2F%2F127.0.0.1%3A59434%2FjfRiHv2G7PA%3D%2F
[        ] Running with unsound null safety
[        ] For more information see https://dart.dev/null-safety/unsound-null-safety
[+6864 ms] [VERBOSE-2:profiler_metrics_ios.mm(203)] Error retrieving thread information: (os/kern) invalid argument
[+6888 ms] flutter: Waiting for biometric authentication.
[  +15 ms] *** Terminating app due to uncaught exception 'NSInvalidArgumentException', reason: 'Non-empty localizedReason must be provided.'
[        ] *** First throw call stack:
[        ] (0x18399f9d8 0x197d25b54 0x1838a24d0 0x1ae82b410 0x1027bddb4 0x1027bccc8 0x103438074 0x102bd7b78 0x102ed6f5c 0x102e7635c 0x102e78a14 0x1839203e0 0x18391ffe4 0x18391f4c4
0x183919850 0x183918ba0 0x19a681598 0x18620a2f4 0x18620f874 0x10278db88 0x1835f7568)
[        ] libc++abi.dylib: terminating with uncaught exception of type NSException
[  +72 ms] Process 474 stopped
[        ] * thread #1, queue = 'com.apple.main-thread', stop reason = signal SIGABRT
[        ]     frame #0: 0x00000001af13684c libsystem_kernel.dylib`__pthread_kill + 8
[        ] libsystem_kernel.dylib`__pthread_kill:
[        ] ->  0x1af13684c <+8>:  b.lo   0x1af136868               ; <+36>
[        ]     0x1af136850 <+12>: stp    x29, x30, [sp, #-0x10]!
[        ]     0x1af136854 <+16>: mov    x29, sp
[        ]     0x1af136858 <+20>: bl     0x1af113f5c               ; cerror_nocancel
[        ] Target 0: (Runner) stopped.
[        ] ios-deploy exited with code 0
[  +63 ms] Service protocol connection closed.
[        ] Lost connection to device.
[   +2 ms] DevFS: Deleting filesystem on the device
(file:///private/var/mobile/Containers/Data/Application/3BE26447-7E2D-4B91-A697-918D964971E0/tmp/local_auth_bugbQutbQ/local_auth_bug/)
[ +254 ms] Ignored error while cleaning up DevFS: TimeoutException after 0:00:00.250000: Future not completed
[   +2 ms] "flutter run" took 63,296ms.
[ +128 ms] ensureAnalyticsSent: 124ms
[   +1 ms] Running shutdown hooks
[        ] Shutdown hook priority 4
[   +1 ms] Shutdown hooks complete
[        ] exiting with code 0
```
The `localizedReason` cannot be empty. 

A good idea can be to manage this exception in a more clear way.

**Comment by TahaTesser** [comment]
Hi @enricobenedos 
Thanks for filing the issue, I can confirm the issue on the simulator. I can't see any indication in Xcode

<details> 
<summary>logs</summary> 

```bash
[ +275 ms] executing: xcrun simctl install BD94A45F-A994-4FC3-BAD9-448D8D2E0F1B
/Users/tahatesser/AndroidStudioProjects/triage-examples/flutter_plugin_examples/local_auth_example/build/ios/iphonesimulator/R
unner.app
[+2475 ms] executing: /usr/bin/plutil -convert json -o -
/Users/tahatesser/AndroidStudioProjects/triage-examples/flutter_plugin_examples/local_auth_example/build/ios/iphonesimulator/R
unner.app/Info.plist
[  +12 ms] Exit code 0 from: /usr/bin/plutil -convert json -o -
/Users/tahatesser/AndroidStudioProjects/triage-examples/flutter_plugin_examples/local_auth_example/build/ios/iphonesimulator/R
unner.app/Info.plist
[        ]
{"CFBundleName":"triage","DTXcode":"1240","DTSDKName":"iphonesimulator14.4","UILaunchStoryboardName":"LaunchScreen","CFBundleI
cons~ipad":{"CFBundlePrimaryIcon":{"CFBundleIconFiles":["AppIcon20x20","AppIcon29x29","AppIcon40x40","AppIcon60x60","AppIcon76
x76","AppIcon83.5x83.5"],"CFBundleIconName":"AppIcon"}},"DTSDKBuild":"18D46","CFBundleDevelopmentRegion":"en","CFBundleVersion
":"1","BuildMachineOSBuild":"20D91","DTPlatformName":"iphonesimulator","CFBundlePackageType":"APPL","UIMainStoryboardFile":"Ma
in","CFBundleSupportedPlatforms":["iPhoneSimulator"],"NSFaceIDUsageDescription":"Why is my app authenticating using face
id?","CFBundleShortVersionString":"1.0.0","CFBundleInfoDictionaryVersion":"6.0","CFBundleExecutable":"Runner","DTCompiler":"co
m.apple.compilers.llvm.clang.1_0","UISupportedInterfaceOrientations~ipad":["UIInterfaceOrientationPortrait","UIInterfaceOrient
ationPortraitUpsideDown","UIInterfaceOrientationLandscapeLeft","UIInterfaceOrientationLandscapeRight"],"MinimumOSVersion":"8.0
","CFBundleIdentifier":"com.nevercode.triage","UIDeviceFamily":[1,2],"DTPlatformVersion":"14.4","CFBundleSignature":"????","CF
BundleIcons":{"CFBundlePrimaryIcon":{"CFBundleIconFiles":["AppIcon20x20","AppIcon29x29","AppIcon40x40","AppIcon60x60"],"CFBund
leIconName":"AppIcon"}},"DTXcodeBuild":"12D4e","LSRequiresIPhoneOS":true,"UISupportedInterfaceOrientations":["UIInterfaceOrien
tationPortrait","UIInterfaceOrientationLandscapeLeft","UIInterfaceOrientationLandscapeRight"],"UIViewControllerBasedStatusBarA
ppearance":false,"NSBonjourServices":["_dartobservatory._tcp"],"DTPlatformBuild":"18D46","NSLocalNetworkUsageDescription":"All
ow Flutter tools on your computer to connect and debug your application. This prompt will not appear on release builds."}
[   +4 ms] executing: xcrun simctl launch BD94A45F-A994-4FC3-BAD9-448D8D2E0F1B com.nevercode.triage --enable-dart-profiling
--enable-checked-mode --verify-entry-points --observatory-port=0
[ +260 ms] com.nevercode.triage: 48981
[        ] Waiting for observatory port to be available...
[ +500 ms] Observatory URL on device: http://127.0.0.1:55613/3hOxk9eh43I=/
[   +8 ms] Caching compiled dill
[ +141 ms] Connecting to service protocol: http://127.0.0.1:55613/3hOxk9eh43I=/
[ +217 ms] Launching a Dart Developer Service (DDS) instance at http://127.0.0.1:0, connecting to VM service at
http://127.0.0.1:55613/3hOxk9eh43I=/.
[ +104 ms] DDS is listening at http://127.0.0.1:55618/MtZi9DKecKg=/.
[ +109 ms] Successfully connected to service protocol: http://127.0.0.1:55613/3hOxk9eh43I=/
[ +819 ms] DevFS: Creating new filesystem on the device (null)
[  +13 ms] DevFS: Created new filesystem on the device
(file:///Users/tahatesser/Library/Developer/CoreSimulator/Devices/BD94A45F-A994-4FC3-BAD9-448D8D2E0F1B/data/Containers/Data/Ap
plication/4B871FC8-2416-4D8F-9AF5-BD77872EA465/tmp/local_auth_exampleBdctnK/local_auth_example/)
[   +8 ms] Updating assets
[  +88 ms] Syncing files to device iPhone 12...
[   +1 ms] <- reset
[        ] Compiling dart to kernel with 0 updated files
[   +5 ms] <- recompile package:triage/main.dart ac65073d-458a-4db3-8f27-333cfbe3e1fc
[        ] <- ac65073d-458a-4db3-8f27-333cfbe3e1fc
[ +158 ms] Updating files.
[        ] DevFS: Sync finished
[        ] Syncing files to device iPhone 12... (completed in 167ms)
[        ] Synced 0.0MB.
[   +1 ms] <- accept
[   +5 ms] Connected to _flutterView/0x7fa1d000ec20.
[   +2 ms] Flutter run key commands.
[   +1 ms] r Hot reload. 🔥🔥🔥
[   +1 ms] R Hot restart.
[   +1 ms] h Repeat this help message.
[        ] d Detach (terminate "flutter run" but leave application running).
[        ] c Clear the screen
[        ] q Quit (terminate the application on the device).
[        ] An Observatory debugger and profiler on iPhone 12 is available at: http://127.0.0.1:55618/MtZi9DKecKg=/
[        ] 
           Flutter DevTools, a Flutter debugger and profiler, on iPhone 12 is available at:
           http://127.0.0.1:55621?uri=http%3A%2F%2F127.0.0.1%3A55618%2FMtZi9DKecKg%3D%2F
[        ] Running with unsound null safety
[        ] For more information see https://dart.dev/null-safety/unsound-null-safety
[+49131 ms] flutter: Waiting for biometric authentication.
[ +207 ms] Service protocol connection closed.
[        ] Lost connection to device.
[   +5 ms] DevFS: Deleting filesystem on the device
(file:///Users/tahatesser/Library/Developer/CoreSimulator/Devices/BD94A45F-A994-4FC3-BAD9-448D8D2E0F1B/data/Containers/Data/Ap
plication/4B871FC8-2416-4D8F-9AF5-BD77872EA465/tmp/local_auth_exampleBdctnK/local_auth_example/)
[ +262 ms] Ignored error while cleaning up DevFS: TimeoutException after 0:00:00.250000: Future not completed
[   +4 ms] "flutter run" took 158,403ms.
[ +176 ms] ensureAnalyticsSent: 171ms
[   +1 ms] Running shutdown hooks
[        ] Shutdown hook priority 4
[   +3 ms] Shutdown hooks complete
[        ] exiting with code 0
```

</details>

Thanks for finding the cause,  I agree

> A good idea can be to manage this exception in a more clear way.



<details> 
<summary>flutter doctor -v</summary> 

```bash
[✓] Flutter (Channel stable, 2.0.1, on macOS 11.2.3 20D91 darwin-x64, locale
    en-GB)
    • Flutter version 2.0.1 at /Users/tahatesser/Code/flutter_stable
    • Framework revision c5a4b4029c (6 days ago), 2021-03-04 09:47:48 -0800
    • Engine revision 40441def69
    • Dart version 2.12.0

[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.3)
    • Android SDK at /Volumes/Extreme/SDK
    • Platform android-30, build-tools 30.0.3
    • ANDROID_HOME = /Volumes/Extreme/SDK
    • Java binary at: /Applications/Android
      Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6915495)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS
    • Xcode at /Volumes/Extreme/Xcode.app/Contents/Developer
    • Xcode 12.4, Build version 12D4e
    • CocoaPods version 1.10.1

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 4.1)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6915495)

[✓] VS Code (version 1.54.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.20.0

[✓] Connected device (5 available)
    • SM M025F (mobile)    • R9ZR205XX0A                          • android-arm
      • Android 10 (API 29)
    • Taha’s iPad (mobile) • 00008020-000255113EE8402E            • ios
      • iOS 14.4.1
    • iPhone 12 (mobile)   • BD94A45F-A994-4FC3-BAD9-448D8D2E0F1B • ios
      • com.apple.CoreSimulator.SimRuntime.iOS-14-4 (simulator)
    • macOS (desktop)      • macos                                • darwin-x64
      • macOS 11.2.3 20D91 darwin-x64
    • Chrome (web)         • chrome                               •
      web-javascript • Google Chrome 89.0.4389.82

• No issues found!
```

</details>

**Comment by enricobenedos** [comment]
Hi @TahaTesser,

I can create a pull request with an "high level" fix adding an `assert(localizedReason.isNotEmpty);` and making some adjunctive tests. At this point the developer can diagnose the failed assertion during the debug process without app crashes.
